tools/reactive-controllers/src/ElementResolution.ts

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2022 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import type { ReactiveController, ReactiveElement } from 'lit';
+export const elementResolverUpdatedSymbol = Symbol('element resolver updated');
+
+export class ElementResolutionController implements ReactiveController {
+    get element(): HTMLElement | null {
+        return this._element;
+    }
+
+    set element(element: HTMLElement | null) {
+        if (element === this.element) return;
+        const previous = this.element;
+        this._element = element;
+        this.host.requestUpdate(elementResolverUpdatedSymbol, previous);
+    }
+
+    private _element: HTMLElement | null = null;
+
+    private host!: ReactiveElement;
+
+    private observer!: MutationObserver;
+
+    get selector(): string {
+        return this._selector;
+    }
+
+    set selector(selector: string) {
+        if (selector === this.selector) return;
+        this.releaseElement();
+        this._selector = selector;
+        this.resolveElement();
+    }
+
+    private _selector = '';
+
+    constructor(
+        host: ReactiveElement,
+        { selector }: { selector: string } = { selector: '' }
+    ) {
+        this.host = host;
+        this.host.addController(this);
+        this.selector = selector;
+        this.observer = new MutationObserver((mutationList) => {
+            let needsResolution = false;
+            mutationList.forEach((mutation) => {
+                if (needsResolution) return;
+                switch (mutation.type) {
+                    case 'childList':
+                        const currentElementRemoved =
+                            this.element &&
+                            [...mutation.removedNodes].includes(this.element);
+                        const matchingElementAdded =
+                            !!this.selector &&
+                            ([...mutation.addedNodes] as HTMLElement[]).some(
+                                (el) => el?.matches?.(this.selector)
+                            );
+                        needsResolution =
+                            needsResolution ||
+                            currentElementRemoved ||
+                            matchingElementAdded;
+                        break;
+                    case 'attributes':
+                        const attributeChangedOnCurrentElement =
```

Comment:
Sometime this talks about `selectors` sometimes it talks about `ids` which are not 100% the same, which is the right one, generic resolution or ID resolution.